### PR TITLE
Add varnish

### PR DESCRIPTION
**Pull request by gquintard**
# Context

This PR is going to compete with #4404, with a smaller scope, but hopefully more maintainability. 

Mainly, the images only target `amd64`, leverage the upstream packages on [packagecloud.io](https://packagecloud.io/varnishcache/), and are debian only at the moment (alpine upport is definitely possible, but I want to validate my approach on this PR first).

Ahead of the checklist:

- I am a Varnish Software employee
- the category for the image is "service" I believe
- I'm not a docker expert, and while I read the guidelines, there may be some mistakes left, I'll fix them as soon as you point them out

# Checklist for Review

**NOTE:** This checklist is intended for the use of the Official Images maintainers both to track the status of your PR and to help inform you and others of where we're at. As such, please leave the "checking" of items to the repository maintainers. If there is a point below for which you would like to provide additional information or note completion, please do so by commenting on the PR. Thanks! (and thanks for staying patient with us :heart:)

-	[x] associated with or contacted upstream?
-	[x] does it fit into one of the common categories? ("service", "language stack", "base distribution")
-	[x] is it reasonably popular, or does it solve a particular use case well?
-	[x] does a [documentation](https://github.com/docker-library/docs/blob/master/README.md) PR exist? (should be reviewed and merged at roughly the same time so that we don't have an empty image page on the Hub for very long)
-	[x] dockerization review for best practices and cache gotchas/improvements (ala [the official review guidelines](https://github.com/docker-library/official-images/blob/master/README.md#review-guidelines))?
-	[x] 2+ dockerization review?
-	[x] existing official images have been considered as a base? (ie, if `foobar` needs Node.js, has `FROM node:...` instead of grabbing `node` via other means been considered?)
-	[x] ~if `FROM scratch`, tarballs only exist in a single commit within the associated history?~
-	[x] passes current tests? any simple new tests that might be appropriate to add? (https://github.com/docker-library/official-images/tree/master/test)

cheers!

### CODE REVIEW

**Comment by gquintard** [comment]
Just force-pushed to include @J0WI's cleaning PR.

**Comment by teohhanhui** [comment]
@gquintard I hope we can reconcile on this. This kind of competition will not get us closer to having an official Docker image.

I've talked to our CEO, and he's okay with us transferring ownership of the `coopTilleuls/docker-varnish` to under the `varnish` organization, provided I'm kept as the maintainer, and a message that says the project is sponsored by our company.

I wish to remind you of the history of the work:

2015–2016: It started at The New York Times Company, but became unmaintained.
2016–2017: I forked the repository at Tripviss, my previous company.
2018–current: We moved the repository under Les-Tilleuls.coop, and we use it in API Platform.

I hope Varnish Software will work with the Docker community, and we don't have to throw away years of hard work that have enabled us to come this close to having an official Docker image.

On a technical point, I understand why you're frustrated with the complicated Dockerfile, but it's best that we stay close to another Docker official image. We stand to gain a lot by learning from the experience of the Docker team and community, not disregarding them.

I hope for a productive outcome for all of us. Thank you.

**Comment by gquintard** [comment]
@teohhanhui, I have to thank you for your continued work on this, and for seeking upstream's feedback.

I respect your stance on #4404, and understand why you are standing your ground, but I obviously disagree with the technical conclusions (reusing upstream packaging work makes more sense to me, and I'd rather have code I trust because I understand it than code adapted from code I was told to trust).
Reading the guidelines and the comments of #4404, I believe this PR hits all the marks, in a way that the original can't.

That being said, I'll state again that I'm no authority on the matter, and I trust the project to referee on this as they have the most experience. If to them this PR is not a sounder base than #4404, I'll scrap it, no worries, but as the topic has been opened for nearly a year, I feel there is gain to offer an alternative (sorry if "competing" felt aggressive) approach.

Regarding ownership, I'm sure we'll figure something out, let's first get an official image in the repository.

All that being said, this is not technical talk, and we should keep that thread clean to ease the maintainers' work. Can you hit me up at guillaume@varni.sh to discuss the details further (and I believe we can even discuss this in French?).

**Comment by tianon** [comment]
Sorry for the delay on getting to this.  I want to preface by saying I appreciate the interest in Varnish, especially directly from an official upstream representative/maintainer (and am always glad to see additional community support and/or willingness to help maintain).

If forced to make a choice here, and with absolutely nothing but respect and appreciate for all parties involved, we're going to go with Varnish to maintain and represent Varnish within the program, but I'd like to think that we could come to some sort of agreement on what's best for the Varnish project without resorting to such a "hammer" solution.

From what I can see, the primary advantage of the approach here is to reuse the existing artifacts that the Varnish project officially publishes, maintains, and supports (which is something we commonly strongly recommend image maintainers consider doing).  The primary advantage I see to #4404 is that it can support a much broader set of architectures.

I guess my first question to try and help mediate here would be whether the Varnish project officially supports those additional architectures to begin with?  To put that another way, if a user on ARM64 or s390x were to file an issue on the Varnish project, would it receive the same level of debugging/care as a user with an issue on x86/amd64?

In other words, does https://packagecloud.io/varnishcache/ only contain `amd64` binaries due to a lack of build infrastructure, or is there more to it than that?

**Comment by gquintard** [comment]
Thanks for the feedback!

We only build the `amd64` package mainly because of a lack of resources and because Varnish targets 64 bits architectures first and foremost. So we will be glad to be able to build other 64 bits architectures, 32 bits are not really interesting.

For a bit more context, I discussed with @teohhanhui  out-of-band and came to an agreement: we can go for an approach like #4404 if, while building the packages, we can `make distcheck`, and in the meantime, we'll use the packagecloud.io packages.

We are very interested in the alpine packages notably, but at the moment the build isn't completely successful so we can't put a stamp on it. But show us a build on a platform, without too much cargo-culting, and we're game.

Does that clarify things a bit?

**Comment by tianon** [comment]
Yeah, for sure -- so it sounds like you're saying we should start with this PR and then revisit the approach in #4404 as a separate conversation on the Varnish-maintained image as further improvement to expand the supported architectures list.  Is that an accurate re-stating?  That would allow for this to grow more organically and makes sure that y'all at Varnish upstream are comfortable support the things that get published here (which I think is important), so :+1: from me.

---

Let's do some Dockerization review! :tada:

1. see https://github.com/docker-library/faq#openpgp--gnupg-keys-and-verification -- TLDR, `COPY gpgkey ...` is hard for users to verify, annoying during review (since the full exported key file shows up in our diffs), etc; I'd recommend instead pulling that key via full fingerprint from a PGP keyserver (both so users can verify it with a fingerprint that's hopefully published on an official Varnish page and so it gets fully verified during the image build)

2. additionally, `apt-key add` is discouraged by the APT maintainers, so I'd recommend doing something with `gpg --batch ... --recv-keys FULL-FINGERPRINT` followed by something like `gpg --batch --export --armor FULL-FINGERPRINT > /etc/apt/trusted.gpg.d/varnish.gpg.asc` (see https://github.com/docker-library/mysql/blob/6b4cb5918a63baaf8c6adacb760f867e034587a6/8.0/Dockerfile#L38-L46 for an example of how that ends up looking in practice, happy to provide others or to provide a sample `Dockerfile` block if that'd be more helpful)

3. regarding https://github.com/varnish/docker-varnish/blob/6a28dbedf5851b51f8fbefeb6ee928d8ae596298/stable/debian/docker-varnish-entrypoint#L7 -- are there any `varnishd` flags or subcommands that don't start with a hyphen?  (if so, maybe there's a technique that could support those too?  see https://github.com/docker-library/docker/blob/fded04bc1ec3920a5960a76e398d7ce5f147f6f7/docker-entrypoint.sh#L9-L13 for an example of what we do in the Docker official image)

**Comment by gquintard** [comment]
Hi @tianon,

Thank you for the review, and sorry for the delay, things have been busy.

I believe the `gpg` stuff is now up to par. I do seldom have a `gpg` error when importing the key, but it seems to be related to my setup and IPv6, not sure how grave it is as the code is pretty much the same as the example you provided.

`varnishd`'s flags all have a hyphen, so no issue here. In the future, it'll be nice to provide a layer of abstraction to configure/run, but for now I think we are good staying close to the metal. Thoughts?

**Comment by yosifkit** [comment]
Thanks @gquintard! I have a couple requests and questions:

- the gpg key should be the full fingerprint rather than just the long id.
- can we add `--keyserver ha.pool.sks-keyservers.net` to the `recv-keys` command?
- `apt-transport-https`: I think it'd be better the leave this installed, that way if someone does an image `FROM varnish` they won't fail on an `apt-get update`.
- `varnish:fresh`: is "fresh" a known alias for the newest varish version? We don't have any other images with this tag name and from my perspective seems redundant with `latest`.

The only other thing I have is a request for a PR to the docs repo, similar to https://github.com/docker-library/docs/pull/1241. This will be used to generate the content of the Docker Hub page (like https://hub.docker.com/_/redis/). Feel free to ping here or in https://github.com/docker-library/docs if you have questions.

**Comment by gquintard** [comment]
* full fingerprint done
* specific key server done
* `apt-transport-https` left behind

Last point if the `fresh` nomenclature. This is specific to varnish (https://varnish-cache.org/releases/index.html), we have:
- fresh: released during the last cycle (releases occur on 15/03 and 15/09)
- lts/supported: a fresh release that got promoted and will be supported for some time
- weekly: promoted every week, as you will have guessed

For each type, like immortals, there can only be one. I can rename `fresh` to `latest`, no problem. It's just that if we add a `weekly` section to the repo, there will be confusion between `latest` and `weekly`, I believe.

But the goal here is to align as closely and as cleanly as possible to the project guidelines, so, you tell me :-)

As it seems we are getting close to merging, I'll start working on the docs. Again, thank you for your time and attention.

**Comment by gquintard** [comment]
@teohhanhui, as you already did a great job on it, do you mind if I reuse https://github.com/docker-library/docs/pull/1241 almost fully (but I'll keep your commits and you as the author)?

**Comment by yosifkit** [comment]
The `fresh` is fine by me. It will be known to your users and that is what it is for. :+1:

And what is hopefully one final comment (from me), you should add `$GNUPGHOME` to the `rm -rf`.

**Comment by gquintard** [comment]
> And what is hopefully one final comment (from me), you should add $GNUPGHOME to the rm -rf.

done

I'll wait for @tianon's feedback on naming, and get working on the docs

**Comment by teohhanhui** [comment]
> @teohhanhui, as you already did a great job on it, do you mind if I reuse docker-library/docs#1241 almost fully (but I'll keep your commits and you as the author)?

@gquintard Feel free to take what you can! :)

**Comment by teohhanhui** [comment]
I think it's better if the directories could be named after the versions, and `fresh` and `stable` just used as tags.

**Comment by tianon** [comment]
I've got no issues with the `fresh` tag given it's meaningful to Varnish users/upstream. :+1:

(You're welcome to ditch `latest` if you _really_ want to be crazy -- you wouldn't be the first, but that's perhaps a bridge too far.)

**Comment by gquintard** [comment]
let's keeps tags as they are then

docs PR is here: https://github.com/docker-library/docs/pull/1519

File: library/varnish [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# this file was generated using https://github.com/varnish/docker-varnish/blob/8312c29a9a7dbd161c291683a77ac209fe7b3cd0/populate.sh
+Maintainers: Guillaume Quintard <guillaume@varni.sh> (@gquintard)
+GitRepo: https://github.com/varnish/docker-varnish.git
+
+Tags: 6.2, 6.2.0-1, 6.2.0, 6, latest, fresh
+Architectures: amd64
+Directory: fresh/debian
+GitCommit: 8312c29a9a7dbd161c291683a77ac209fe7b3cd0
```
Comment (yosifkit) [inline library/varnish]: One more push to get these to https://github.com/varnish/docker-varnish/commit/82ba69727be4e78de5a3e07a63e8b62d92ab958d

**Comment by yosifkit** [comment]
<details>
<summary>Diff:</summary>

```diff
diff --git a/_bashbrew-arches b/_bashbrew-arches
index e69de29..2d57ca0 100644
--- a/_bashbrew-arches
+++ b/_bashbrew-arches
@@ -0,0 +1,2 @@
+varnish:fresh @ amd64
+varnish:stable @ amd64
diff --git a/_bashbrew-list b/_bashbrew-list
index e69de29..b174832 100644
--- a/_bashbrew-list
+++ b/_bashbrew-list
@@ -0,0 +1,10 @@
+varnish:6
+varnish:6.0
+varnish:6.0.3
+varnish:6.0.3-1
+varnish:6.2
+varnish:6.2.0
+varnish:6.2.0-1
+varnish:fresh
+varnish:latest
+varnish:stable
diff --git a/varnish_fresh/Dockerfile b/varnish_fresh/Dockerfile
new file mode 100644
index 0000000..2043e10
--- /dev/null
+++ b/varnish_fresh/Dockerfile
@@ -0,0 +1,31 @@
+FROM debian:stretch-slim
+
+ENV VARNISH_VERSION 6.2.0-1~stretch
+
+RUN set -ex; \
+	fetchDeps=" \
+		ca-certificates \
+		dirmngr \
+		gnupg \
+	"; \
+	apt-get update; \
+	apt-get install -y --no-install-recommends apt-transport-https $fetchDeps; \
+	key=B54813B54CA95257D3590B3F1B0096460868C7A9; \
+	export GNUPGHOME="$(mktemp -d)"; \
+	gpg --batch --keyserver http://ha.pool.sks-keyservers.net/ --recv-keys $key; \
+	gpg --batch --export export $key > /etc/apt/trusted.gpg.d/varnish.gpg; \
+	gpgconf --kill all; \
+	rm -rf $GNUPGHOME; \
+	echo deb https://packagecloud.io/varnishcache/varnish62/debian/ stretch main > /etc/apt/sources.list.d/varnish.list; \
+	apt-get update; \
+	apt-get install -y --no-install-recommends varnish=$VARNISH_VERSION; \
+	apt-get purge -y --auto-remove -o APT::AutoRemove::RecommendsImportant=false $fetchDeps; \
+	rm -rf /var/lib/apt/lists/*
+
+WORKDIR /etc/varnish
+
+COPY docker-varnish-entrypoint /usr/local/bin/
+ENTRYPOINT ["docker-varnish-entrypoint"]
+
+EXPOSE 80
+CMD ["varnishd", "-F", "-f", "/etc/varnish/default.vcl"]
diff --git a/varnish_fresh/docker-varnish-entrypoint b/varnish_fresh/docker-varnish-entrypoint
new file mode 100755
index 0000000..bde6e74
--- /dev/null
+++ b/varnish_fresh/docker-varnish-entrypoint
@@ -0,0 +1,12 @@
+#!/bin/sh
+set -e
+
+
+# this if will check if the first argument is a flag
+# but only works if all arguments require a hyphenated flag
+# -v; -SL; -f arg; etc will work, but not arg1 arg2
+if [ "$#" -eq 0 ] || [ "${1#-}" != "$1" ]; then
+    set -- varnishd "$@"
+fi
+
+exec "$@"
diff --git a/varnish_stable/Dockerfile b/varnish_stable/Dockerfile
new file mode 100644
index 0000000..66aafca
--- /dev/null
+++ b/varnish_stable/Dockerfile
@@ -0,0 +1,31 @@
+FROM debian:stretch-slim
+
+ENV VARNISH_VERSION 6.0.3-1~stretch
+
+RUN set -ex; \
+	fetchDeps=" \
+		ca-certificates \
+		dirmngr \
+		gnupg \
+	"; \
+	apt-get update; \
+	apt-get install -y --no-install-recommends apt-transport-https $fetchDeps; \
+	key=48D81A24CB0456F5D59431D94CFCFD6BA750EDCD; \
+	export GNUPGHOME="$(mktemp -d)"; \
+	gpg --batch --keyserver http://ha.pool.sks-keyservers.net/ --recv-keys $key; \
+	gpg --batch --export export $key > /etc/apt/trusted.gpg.d/varnish.gpg; \
+	gpgconf --kill all; \
+	rm -rf $GNUPGHOME; \
+	echo deb https://packagecloud.io/varnishcache/varnish60lts/debian/ stretch main > /etc/apt/sources.list.d/varnish.list; \
+	apt-get update; \
+	apt-get install -y --no-install-recommends varnish=$VARNISH_VERSION; \
+	apt-get purge -y --auto-remove -o APT::AutoRemove::RecommendsImportant=false $fetchDeps; \
+	rm -rf /var/lib/apt/lists/*
+
+WORKDIR /etc/varnish
+
+COPY docker-varnish-entrypoint /usr/local/bin/
+ENTRYPOINT ["docker-varnish-entrypoint"]
+
+EXPOSE 80
+CMD ["varnishd", "-F", "-f", "/etc/varnish/default.vcl"]
diff --git a/varnish_stable/docker-varnish-entrypoint b/varnish_stable/docker-varnish-entrypoint
new file mode 100755
index 0000000..bde6e74
--- /dev/null
+++ b/varnish_stable/docker-varnish-entrypoint
@@ -0,0 +1,12 @@
+#!/bin/sh
+set -e
+
+
+# this if will check if the first argument is a flag
+# but only works if all arguments require a hyphenated flag
+# -v; -SL; -f arg; etc will work, but not arg1 arg2
+if [ "$#" -eq 0 ] || [ "${1#-}" != "$1" ]; then
+    set -- varnishd "$@"
+fi
+
+exec "$@"
```

</details>

**Comment by yosifkit** [comment]
LGTM!

Build test of #5855; a5239d43; `amd64` (`varnish`):

```console

$ bashbrew build varnish:6.2
Building bashbrew/cache:0cc78a5a55cb3b88c412b7e00df7960e83ed30fd1dbb26f48e0ffe2d30c77aad (varnish:6.2)
Tagging varnish:6.2
Tagging varnish:6.2.0-1
Tagging varnish:6.2.0
Tagging varnish:6
Tagging varnish:latest
Tagging varnish:fresh

$ test/run.sh varnish:6.2
testing varnish:6.2
	'utc' [1/4]...passed
	'cve-2014--shellshock' [2/4]...passed
	'no-hard-coded-passwords' [3/4]...passed
	'override-cmd' [4/4]...passed


$ bashbrew build varnish:6.0
Building bashbrew/cache:2f36f1118638f3d1fd06d7d71e5514486619e09585dd426b1dcfdb565c081182 (varnish:6.0)
Tagging varnish:6.0
Tagging varnish:6.0.3-1
Tagging varnish:6.0.3
Tagging varnish:stable

$ test/run.sh varnish:6.0
testing varnish:6.0
	'utc' [1/4]...passed
	'cve-2014--shellshock' [2/4]...passed
	'no-hard-coded-passwords' [3/4]...passed
	'override-cmd' [4/4]...passed

```

**Comment by tianon** [comment]
LGTM :+1:

**Pull request merged**